### PR TITLE
Exporter APS Authentication

### PR DESCRIPTION
**Pull request by PepperLola**
Changes:
- Palette to get authorization code for Autodesk account
- Fetches code and refreshes automatically once expired
- Shows first name in a new tab in the main exporter panel (temporary)

Testing:
- Make sure the server is running (same branch name as this pr)
- Sign in with APS button
- Check main panel to see if your name shows up in a third tab
- Make sure login palette automatically closes after signing in
- Make sure reopening palette prompts for login again
- Make sure closing palette manually and reopening prompts for login again
  - Might have to enter email but not password to ensure that it's resetting

> [!WARNING]
> I haven't yet tested refreshing the token once it's expired. The web request works though and gets a different access token than before.

> [!IMPORTANT]
> This branch must be tested while running the server through the Docker container, because some authorization logic happens on the server side.

### CODE REVIEW

**Review by BrandonPacewic — CHANGES_REQUESTED** [review]
I get the following error upon attempt to open the APS panel.

image

Here is the log file:

```
2024-06-28 10:03:56,954 - synthesis.Toolbar.getNewPanel - DEBUG - Created Panel Synthesis_synthesis_tooltab in Toolbar ToolsTab
2024-06-28 10:03:56,972 - synthesis.HUI.HButton - INFO - Created Button SynthesisExporter_synthesis in Panel Synthesis_synthesis_tooltab
2024-06-28 10:03:56,989 - synthesis.HUI.HButton - INFO - Created Button APS_synthesis in Panel Synthesis_synthesis_tooltab
2024-06-28 10:04:19,260 - synthesis - ERROR - Command executed failed: Traceback (most recent call last):
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/urllib/request.py", line 1348, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/http/client.py", line 1282, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/http/client.py", line 1328, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/http/client.py", line 1277, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/http/client.py", line 1037, in _send_output
    self.send(msg)
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/http/client.py", line 975, in send
    self.connect()
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/http/client.py", line 941, in connect
    self.sock = self._create_connection(
                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/socket.py", line 851, in create_connection
    raise exceptions[0]
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/socket.py", line 836, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 61] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/brandon/Documents/Repos/synthesis/exporter/SynthesisFusionAddin/src/UI/ShowAPSAuthCommand.py", line 32, in notify
    challenge = getCodeChallenge()
                ^^^^^^^^^^^^^^^^^^
  File "/Users/brandon/Documents/Repos/synthesis/exporter/SynthesisFusionAddin/src/APS/APS.py", line 64, in getCodeChallenge
    res = urllib.request.urlopen(endpoint)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/urllib/request.py", line 519, in open
    response = self._open(req, data)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/urllib/request.py", line 536, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/urllib/request.py", line 496, in _call_chain
    result = func(*args)
             ^^^^^^^^^^^
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/urllib/request.py", line 1377, in http_open
    return self.do_open(http.client.HTTPConnection, req)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/urllib/request.py", line 1351, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [Errno 61] Connection refused>

2024-06-28 10:07:02,016 - synthesis - ERROR - Command executed failed: Traceback (most recent call last):
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/urllib/request.py", line 1348, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/http/client.py", line 1282, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/http/client.py", line 1328, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/http/client.py", line 1277, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/http/client.py", line 1037, in _send_output
    self.send(msg)
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/http/client.py", line 975, in send
    self.connect()
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/http/client.py", line 941, in connect
    self.sock = self._create_connection(
                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/socket.py", line 851, in create_connection
    raise exceptions[0]
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/socket.py", line 836, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 61] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/brandon/Documents/Repos/synthesis/exporter/SynthesisFusionAddin/src/UI/ShowAPSAuthCommand.py", line 32, in notify
    challenge = getCodeChallenge()
                ^^^^^^^^^^^^^^^^^^
  File "/Users/brandon/Documents/Repos/synthesis/exporter/SynthesisFusionAddin/src/APS/APS.py", line 64, in getCodeChallenge
    res = urllib.request.urlopen(endpoint)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/urllib/request.py", line 519, in open
    response = self._open(req, data)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/urllib/request.py", line 536, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/urllib/request.py", line 496, in _call_chain
    result = func(*args)
             ^^^^^^^^^^^
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/urllib/request.py", line 1377, in http_open
    return self.do_open(http.client.HTTPConnection, req)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brandon/Library/Application Support/Autodesk/webdeploy/production/6997599ab67b7e881f25c2311930a09bc5ac662f/Autodesk Fusion 360.app/Contents/Frameworks/Python.framework/Versions/Current/lib/python3.11/urllib/request.py", line 1351, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [Errno 61] Connection refused>

```

**Review by BrandonPacewic — APPROVED** [review]
Seems to work...

The security workflow is getting real mad at us with this one, not sure if there is anything we can do to fix that though.

**Comment by HunterBarclay** [comment]
@PepperLola Something to note: I forgot to switch the server branch over so I ran into an issue. That being said, after getting the issue, there isn't any way for me to reset the aps login process without restarting the addin. Could we have to detect an error and then automatically reset?

**Comment by HunterBarclay** [comment]
@PepperLola I get this page in the palette, and then nothing further:

image

**Comment by HunterBarclay** [comment]
It does appear to successfully login, however.

image

File: exporter/SynthesisFusionAddin/src/APS/APS.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    if curr_time >= APS_AUTH.expires_at:
+        refreshAuthToken()
+    if APS_USER_INFO is None:
+        loadUserInfo()
+    return APS_AUTH
+
+
+def convertAuthToken(code: str):
+    global APS_AUTH
+    authUrl = f'http://localhost:3003/api/aps/code/?code={code}&redirect_uri={urllib.parse.quote_plus("http://localhost:3003/api/aps/exporter/")}'
```
Comment (azaleacolburn) [inline exporter/SynthesisFusionAddin/src/APS/APS.py]: These will have to be changed to the actually URL eventually, right? A constant might be helpful for this.

**Review by azaleacolburn — APPROVED** [review]
I know I'm super late to the party, but everything worked for me.

**Pull request merged**